### PR TITLE
[Mono] Turn Samples into Functional Tests

### PR DESCRIPTION
**Pull request by MaximLipnin**
Related issue: https://github.com/dotnet/runtime/issues/43865

The idea is to copy existing [mono samples](https://github.com/dotnet/runtime/tree/master/src/mono/netcore/sample) to a particular place within runtime/src/libraries/ directory and tune them to re-use library test infra like it's done for other functional tests.

As a first step, I'll try to do it with iOS sample. Any ideas on how to do it better/more correct are appreciated.

### CODE REVIEW

**Comment by jkotas** [comment]
- Could you please add README.md that describes what these samples are for and how to use them? 
- Do we expect customers to make copy of these samples to create their own apps? Do we provide any guarantees that the sample copied this way will continue to work?
- Would https://github.com/dotnet/runtime/tree/master/src/tests be a more logical place for this?

**Comment by MaximLipnin** [comment]
@jkotas Thank you for a quick look.

> * Could you please add README.md that describes what these samples are for and how to use them?

Do you mean readme for existing samples?

> * Do we expect customers to make copy of these samples to create their own apps? Do we provide any guarantees that the sample copied this way will continue to work?

As far as I know those sample are used for showing off the platforms we support. Also they are useful for some debugging purposes. I don't know if the samples are supposed to be copied to create new apps.

> * Would https://github.com/dotnet/runtime/tree/master/src/tests be a more logical place for this?

I mostly worked with library tests so that's why I started here. If the location you mentioned is better then I'll consider that option.

File: src/libraries/Mono.Samples/Directory.Build.props [hunk 1]
```diff
@@ -0,0 +1,8 @@
+<Project>
+  <PropertyGroup>
+    <RunAnalyzers>false</RunAnalyzers>
+    <TreatWarningsAsErrors>false</TreatWarningsAsErrors>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
```
Comment (lambdageek) [inline src/libraries/Mono.Samples/Directory.Build.props:5]: We're still keeping the samples under `src/mono/netcore/samples`, right? (I think we should keep them, personally - they're part of the runtime team's devloop for investigating issues)

The new samples are are for CI, and we dont' expect to tweak them often?  In that case they don't need these properties - the CI samples should compile cleanly, and they don't need any unsafe blocks.

```suggestion
```

File: src/libraries/Mono.Samples/tests/Functional/iOS/Makefile [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+runtimepack:
+	../../../../.././build.sh Mono+Libs -os iOS -arch $(MONO_ARCH) -c $(MONO_CONFIG)
+
+run: clean appbuilder
+	$(DOTNET) publish -c $(MONO_CONFIG) /p:TargetArchitecture=$(MONO_ARCH) \
+	/p:UseLLVM=$(USE_LLVM) /p:UseAotForSimulator=true
+
+clean:
+	rm -rf bin
```
Comment (lambdageek) [inline src/libraries/Mono.Samples/tests/Functional/iOS/Makefile:21]:
```suggestion
```

We don't need this Makefile. it's not used by CI.

File: src/libraries/Mono.Samples/Directory.Build.props (hunk 1 above)
Comment (MaximLipnin) [reply src/libraries/Mono.Samples/Directory.Build.props]: Thanks!

File: src/libraries/Mono.Samples/tests/Functional/iOS/Makefile (hunk 2 above)
Comment (MaximLipnin) [reply src/libraries/Mono.Samples/tests/Functional/iOS/Makefile]: Thanks!

**Comment by steveisok** [comment]
@jkotas These are no longer samples.  The intent is to run them through AOT/Interp/JIT to give us a good sense that a barebones app is working.

File: src/libraries/Mono.Samples/Directory.Build.props (hunk 1 above)
Comment (steveisok) [reply src/libraries/Mono.Samples/Directory.Build.props]: Ideally, I'd like this new spot to serve both purposes.  The goal for your inner dev loop scenario is to make them more reliable.  They had a tendency to break and we sometimes wouldn't know until later.

**Comment by marek-safar** [comment]
> Do we expect customers to make copy of these samples to create their own apps?

No, these are specific tests which verify that MonoVM based runtime packs work as expected. I don't think CoreCLR has any such tests otherwise it'd be nice to share the logic.

> Would https://github.com/dotnet/runtime/tree/master/src/tests be a more logical place for this?

Yes but apparently there is no infrastructure yet in place to support having them there.

**Comment by jkotas** [comment]
> These are no longer samples. The intent is to run them through AOT/Interp/JIT to give us a good sense that a barebones app is working.

Then the directory should not be called Samples. The best name I can think of is `IntegrationTests`.

I do not think the name should be Mono-specific. You can imagine having e.g. ASP.NET barebone app inside this set of tests too.

File: src/libraries/IntegrationTests/tests/iOS/iOS.Integration.Tests.csproj [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    <OutputType>Exe</OutputType>
+    <OutputPath>bin</OutputPath>
+    <DebugType>Portable</DebugType>
+    <TargetFramework>$(NetCoreAppCurrent)</TargetFramework>
+    <TargetOS>iOS</TargetOS>
+    <MicrosoftNetCoreAppRuntimePackDir>$(ArtifactsBinDir)microsoft.netcore.app.runtime.ios-$(TargetArchitecture)\$(Configuration)\runtimes\ios-$(TargetArchitecture)\</MicrosoftNetCoreAppRuntimePackDir>
+    <EnableTargetingPackDownload>false</EnableTargetingPackDownload>
+    <RuntimeIdentifier>ios-$(TargetArchitecture)</RuntimeIdentifier>
+    <PublishTrimmed>true</PublishTrimmed>
+    <_TrimmerDefaultAction>link</_TrimmerDefaultAction>
```
Comment (marek-safar) [inline src/libraries/IntegrationTests/tests/iOS/iOS.Integration.Tests.csproj:12]:
```suggestion
	<TrimMode>link</TrimMode>
```

File: src/libraries/IntegrationTests/tests/iOS/iOS.Integration.Tests.csproj [hunk 4]
```diff
@@ -0,0 +1,88 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Comment (marek-safar) [inline src/libraries/IntegrationTests/tests/iOS/iOS.Integration.Tests.csproj:1]: The name needs to be better. Is this device or simulator? Is this interpreter, aot, mixed or what mode?

File: src/libraries/IntegrationTests/tests/iOS/Program.cs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        const string msg = "Hello World!\n.NET 5.0";
+        for (int i = 0; i < msg.Length; i++)
+        {
+            // a kind of an animation
+            ios_set_text(msg.Substring(0, i + 1));
+            await Task.Delay(100);
+        }
+
+        Console.WriteLine("Done!");
+        await Task.Delay(-1);
```
Comment (marek-safar) [inline src/libraries/IntegrationTests/tests/iOS/Program.cs:51]: This needs to check some error codes

File: src/libraries/IntegrationTests/tests/iOS/iOS.Integration.Tests.csproj (hunk 4 above)
Comment (MaximLipnin) [reply src/libraries/IntegrationTests/tests/iOS/iOS.Integration.Tests.csproj:1]: I guess the original sample is about AOT on simulator. So the name could be  `src/libraries/IntegrationTests/tests/iOS/iOS.Simulator.Aot.Integration.Tests.csproj`

File: src/libraries/IntegrationTests/tests/iOS/iOS.Integration.Tests.csproj (hunk 3 above)
Comment (MaximLipnin) [reply src/libraries/IntegrationTests/tests/iOS/iOS.Integration.Tests.csproj]: Done, thanks!

File: src/libraries/IntegrationTests/tests/iOS/iOS.Integration.Tests.csproj (hunk 4 above)
Comment (steveisok) [reply src/libraries/IntegrationTests/tests/iOS/iOS.Integration.Tests.csproj:1]: It was either or as `$(UseAotForSimulator)` made that determination.

File: src/libraries/IntegrationTests/tests/iOS/Program.cs (hunk 5 above)
Comment (vargaz) [reply src/libraries/IntegrationTests/tests/iOS/Program.cs:51]: Why is there a 'tests' directory under 'IntegrationTests' ?

Comment (MaximLipnin) [reply src/libraries/IntegrationTests/tests/iOS/Program.cs:51]: This follows the structure, which other library directories have, in order to be included in https://github.com/dotnet/runtime/blob/65d3e2843e89a09943f47d096adc6fe98195f18c/src/libraries/tests.proj#L168

File: src/libraries/IntegrationTests/tests/iOS/iOS.Integration.Tests.csproj (hunk 4 above)
Comment (MaximLipnin) [reply src/libraries/IntegrationTests/tests/iOS/iOS.Integration.Tests.csproj:1]: Done

**Comment by MaximLipnin** [comment]
Closing it in favor of https://github.com/dotnet/runtime/pull/44462

**Pull request closed without merging**